src/main/java/org/folio/linked/data/integration/kafka/sender/search/WorkCreateMessageSender.java

```diff
@@ -0,0 +1,85 @@
+package org.folio.linked.data.integration.kafka.sender.search;
+
+import static java.lang.String.format;
+import static org.folio.ld.dictionary.ResourceTypeDictionary.WORK;
+import static org.folio.linked.data.util.BibframeUtils.extractWork;
+import static org.folio.linked.data.util.Constants.FOLIO_PROFILE;
+import static org.folio.linked.data.util.Constants.NOT_INDEXED;
+import static org.folio.linked.data.util.Constants.SEARCH_RESOURCE_NAME;
+import static org.folio.search.domain.dto.ResourceIndexEventType.CREATE;
+
+import java.util.List;
+import java.util.Optional;
+import java.util.UUID;
+import lombok.RequiredArgsConstructor;
+import lombok.extern.log4j.Log4j2;
+import org.folio.linked.data.integration.kafka.sender.CreateMessageSender;
+import org.folio.linked.data.mapper.kafka.search.KafkaSearchMessageMapper;
+import org.folio.linked.data.model.entity.Resource;
+import org.folio.linked.data.model.entity.event.ResourceIndexedEvent;
+import org.folio.linked.data.repo.ResourceRepository;
+import org.folio.marc4ld.util.ResourceKind;
+import org.folio.search.domain.dto.LinkedDataWork;
+import org.folio.search.domain.dto.ResourceIndexEvent;
+import org.folio.spring.tools.kafka.FolioMessageProducer;
+import org.springframework.beans.factory.annotation.Qualifier;
+import org.springframework.context.ApplicationEventPublisher;
+import org.springframework.context.annotation.Profile;
+import org.springframework.stereotype.Service;
+
+@Log4j2
+@Service
+@Profile(FOLIO_PROFILE)
+@RequiredArgsConstructor
+public class WorkCreateMessageSender implements CreateMessageSender {
+
+  @Qualifier("bibliographicMessageProducer")
+  private final FolioMessageProducer<ResourceIndexEvent> bibliographicMessageProducer;
+  private final KafkaSearchMessageMapper<LinkedDataWork> searchBibliographicMessageMapper;
+  private final ApplicationEventPublisher eventPublisher;
+  private final ResourceRepository resourceRepository;
+
+  @Override
+  public boolean test(Resource resource) {
+    return ResourceKind.BIBLIOGRAPHIC
+      .stream()
+      .anyMatch(resource::isOfType);
+  }
+
+  @Override
+  public void accept(Resource resource) {
+    if (resource.isOfType(WORK)) {
+      sendWorkCreated(resource);
+    } else {
+      extractWork(resource)
+        .map(Resource::getId)
+        .map(resourceRepository::getReferenceById)
+        .ifPresentOrElse(this::sendWorkCreated,
+          () -> log.warn(format(NOT_INDEXED, resource.getId(), "created")));
+    }
+  }
+
+  private void sendWorkCreated(Resource resource) {
+    searchBibliographicMessageMapper.toIndex(resource, CREATE)
+      .map(this::getCreateIndexEvent)
+      .ifPresent(indexEvent -> {
+        bibliographicMessageProducer.sendMessages(List.of(indexEvent));
+        publishIndexEvent(resource);
+      });
+  }
+
+  private ResourceIndexEvent getCreateIndexEvent(LinkedDataWork linkedDataWork) {
+    return new ResourceIndexEvent()
+      .id(UUID.randomUUID().toString())
```

Comment:
Generic comment for all the kafka messages we send out.

In production, the KAFKA cluster is going to be partitioned for horizontal scaling. If we use a random ID for ResourceIndexEvent, kafka will use the random ID as the message key (see `producer.setKeyMapper(ResourceIndexEvent::getId);` in the producer configurations) and hence the messages generated for the same resource will end up in different partitions. (For example, create message for a resource will go to partition1, an update message for the same resource may go to a partition 2, delete message may go to a partition 3). 

If we use resource's hash as the id, then kafka will guarantee that all messages generated for a resource will end up in the same partition's log, preserving the order. 

If you want, I can create a separate tech debt ticket for using the resource hash as the ID of all kafka messages.

Reply:
It's our general logic was implemented before
It can be too many changes and side effects (maybe not)
A separate ticket will be better, just in case

Reply:
Agree, it should be separated ticket, since current pr contains refactoring with no changes in logic